crucible-llvm/src/Lang/Crucible/LLVM/SimpleLoopFixpoint.hs

```diff
@@ -0,0 +1,748 @@
+
+
+filterSubstitution ::
+  C.IsSymInterface sym =>
+  sym ->
+  MapF (W4.SymExpr sym) (FixpointEntry sym) ->
+  MapF (W4.SymExpr sym) (FixpointEntry sym)
+filterSubstitution sym substitution =
+  -- TODO: fixpoint
+  let uninterp_constants = foldMapF
+        (Set.map (C.mapSome $ W4.varExpr sym) . W4.exprUninterpConstants sym . bodyValue)
+        substitution
+  in
+  MapF.filterWithKey (\variable _entry -> Set.member (C.Some variable) uninterp_constants) substitution
+
+uninterpretedConstantEqualitySubstitution ::
+  forall sym .
+  C.IsSymInterface sym =>
+  sym ->
+  MapF (W4.SymExpr sym) (FixpointEntry sym) ->
+  (MapF (W4.SymExpr sym) (FixpointEntry sym), MapF (W4.SymExpr sym) (W4.SymExpr sym))
+uninterpretedConstantEqualitySubstitution _sym substitution =
+  let reverse_substitution = MapF.foldlWithKey'
+        (\accumulator variable entry -> MapF.insert entry variable accumulator)
+        MapF.empty
+        substitution
+      uninterpreted_constant_substitution = fmapF
+        (\entry -> fromJust $ MapF.lookup entry reverse_substitution)
+        substitution
+      normal_substitution = MapF.filterWithKey
+        (\variable _entry ->
+          Just Refl == W4.testEquality variable (fromJust $ MapF.lookup variable uninterpreted_constant_substitution))
+        substitution
+  in
+  (normal_substitution, uninterpreted_constant_substitution)
+
+
+userSymbol' :: String -> W4.SolverSymbol
+userSymbol' = fromRight (C.panic "SimpleLoopFixpoint.userSymbol'" []) . W4.userSymbol
+
+
+simpleLoopFixpoint ::
+  forall sym ext p rtp blocks init ret .
+  (C.IsSymInterface sym, C.HasLLVMAnn sym, ?memOpts :: C.MemOptions) =>
+  sym ->
+  C.CFG ext blocks init ret ->
+  C.GlobalVar C.Mem ->
+  (MapF (W4.SymExpr sym) (FixpointEntry sym) -> W4.Pred sym -> IO (MapF (W4.SymExpr sym) (W4.SymExpr sym), W4.Pred sym)) ->
+  IO (C.ExecutionFeature p sym ext rtp)
+simpleLoopFixpoint sym cfg@C.CFG{..} mem_var fixpoint_func = do
+  let ?ptrWidth = knownNat @64
+
+  verbSetting <- W4.getOptionSetting W4.verbosity $ W4.getConfiguration sym
+  verb <- fromInteger <$> W4.getOpt verbSetting
+
+  let flattenWTOComponent = \case
+        C.SCC C.SCCData{..} ->  wtoHead : concatMap flattenWTOComponent wtoComps
+        C.Vertex v -> [v]
+  let loop_map = Map.fromList $ mapMaybe
+        (\case
+          C.SCC C.SCCData{..} -> Just (wtoHead, wtoHead : concatMap flattenWTOComponent wtoComps)
+          C.Vertex{} -> Nothing)
+        (C.cfgWeakTopologicalOrdering cfg)
+
+  fixpoint_stuff_ref <- newIORef @(FixpointStuff sym blocks) $ FixpointStuff
+    { fixpointStatus = BeforeFixpoint
+    , fixpointBlockId = undefined
+    , fixpointAssumptionFrameIdentifier = undefined
+    , fixpointSubstitution = undefined
+    , fixpointRegMap = undefined
+    , fixpointMemSubstitution = undefined
+    , fixpointLoopCondition = undefined
+    , fixpointLoopIndexBound = undefined
```

Comment:
We might prefer `error` or `panic` here... at least to let us know where these came from if anyone ever refactors this and introduces a mistake that uses these before they get defined... or are they used? I'm assuming they are filled in eventually

Reply:
It looks like the `fixpointStatus` is being used to encode a state machine here.  I think it would be better to refactor the `FixpointStuff` type with different constructors for the cases so we don't need to have a bunch of fields that are `undefined` until we are ready to fill them in.